File: go/day06/main.go
package main

import (
	"aoc2024/utils"
	"fmt"
)

type Position struct {
	row, col int
}

type State struct {
	pos       Position
	direction string
}

var directions = map[string]Position{
	"^": {-1, 0},
	">": {0, 1},
	"v": {1, 0},
	"<": {0, -1},
}

func getNextDirection(direction string) string {
	switch direction {
	case "^":
		return ">"
	case ">":
		return "v"
	case "v":
		return "<"
	case "<":
		return "^"
	default:
		return ""
	}
}

func getNextPosition(pos Position, direction string) Position {
	delta := directions[direction]
	return Position{pos.row + delta.row, pos.col + delta.col}
}

func buildGrid(path string) [][]string {
	lines := utils.ReadInput(path)
	grid := make([][]string, len(lines))
	for i, line := range lines {
		grid[i] = make([]string, len(line))
		for j, ch := range line {
			grid[i][j] = string(ch)
		}
	}
	return grid
}

func findGuard(grid [][]string) (string, Position) {
	for i := range grid {
		for j := range grid[i] {
			cell := grid[i][j]
			if _, ok := directions[cell]; ok {
				return cell, Position{i, j}
			}
		}
	}
	return "", Position{}
}

func isValidPosition(grid [][]string, pos Position) bool {
	return pos.row >= 0 && pos.row < len(grid) && pos.col >= 0 && pos.col < len(grid[0])
}

func isObstacle(grid [][]string, pos Position) bool {
	return grid[pos.row][pos.col] == "#"
}

func simulateMovement(grid [][]string, direction string, pos Position, visited map[Position]bool) map[Position]bool {
	nextPos := getNextPosition(pos, direction)

	if !isValidPosition(grid, nextPos) {
		return visited
	}

	if isObstacle(grid, nextPos) {
		newDirection := getNextDirection(direction)
		return simulateMovement(grid, newDirection, pos, visited)
	}

	visited[nextPos] = true
	return simulateMovement(grid, direction, nextPos, visited)
}

func createsLoop(grid [][]string, direction string, startPos Position, obstaclePos Position, visitedStates map[State]bool) bool {
	nextPos := getNextPosition(startPos, direction)
	state := State{startPos, direction}

	if visitedStates[state] {
		return true
	}

	if !isValidPosition(grid, nextPos) {
		return false
	}

	visitedStates[state] = true

	if nextPos == obstaclePos || isObstacle(grid, nextPos) {
		newDirection := getNextDirection(direction)
		return createsLoop(grid, newDirection, startPos, obstaclePos, visitedStates)
	}

	return createsLoop(grid, direction, nextPos, obstaclePos, visitedStates)
}

func Part1() {
	grid := buildGrid("day06/input.txt")
	direction, pos := findGuard(grid)
	visited := make(map[Position]bool)
	visited[pos] = true

	result := simulateMovement(grid, direction, pos, visited)
	fmt.Println("Part 1", len(result))
}

func Part2() {
	grid := buildGrid("day06/input.txt")
	initialDirection, initialPos := findGuard(grid)

	total := 0
	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid[0]); col++ {
			pos := Position{row, col}
			if pos != initialPos && !isObstacle(grid, pos) {
				visitedStates := make(map[State]bool)
				if createsLoop(grid, initialDirection, initialPos, pos, visitedStates) {
					total++
				}
			}
		}
	}

	fmt.Println("Part 2", total)
}

func main() {
	Part1()
	Part2()
}

File: rust/day06/src/main.rs
use std::collections::HashSet;
use utils;

#[derive(Clone, Copy, Hash, Eq, PartialEq)]
struct Position {
    row: i32,
    col: i32,
}

impl Position {
    fn next(&self, direction: char) -> Position {
        let delta = match direction {
            '^' => (-1, 0),
            '>' => (0, 1),
            'v' => (1, 0),
            '<' => (0, -1),
            _ => (0, 0),
        };
        Position {
            row: self.row + delta.0,
            col: self.col + delta.1,
        }
    }

    fn is_valid(&self, rows: i32, cols: i32) -> bool {
        self.row >= 0 && self.row < rows && self.col >= 0 && self.col < cols
    }
}

#[derive(Hash, Eq, PartialEq)]
struct State {
    pos: Position,
    direction: char,
}

const DIRECTIONS: [char; 4] = ['^', '>', 'v', '<'];
const NEXT_DIRECTION: [(char, char); 4] = [('^', '>'), ('>', 'v'), ('v', '<'), ('<', '^')];

fn get_next_direction(direction: char) -> char {
    NEXT_DIRECTION
        .iter()
        .find(|&&(curr, _)| curr == direction)
        .map_or(' ', |&(_, next)| next)
}

fn build_grid(path: &str) -> Vec<Vec<char>> {
    utils::read_input(path)
        .lines()
        .map(|line| line.chars().collect())
        .collect()
}

fn find_guard(grid: &[Vec<char>]) -> (char, Position) {
    for (i, row) in grid.iter().enumerate() {
        for (j, &cell) in row.iter().enumerate() {
            if DIRECTIONS.contains(&cell) {
                return (
                    cell,
                    Position {
                        row: i as i32,
                        col: j as i32,
                    },
                );
            }
        }
    }
    (' ', Position { row: 0, col: 0 })
}

fn simulate_movement(
    grid: &[Vec<char>],
    initial_direction: char,
    initial_pos: Position,
    visited: &mut HashSet<Position>,
) {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;
    let mut current_pos = initial_pos;
    let mut current_direction = initial_direction;

    loop {
        let next_pos = current_pos.next(current_direction);

        if !next_pos.is_valid(rows, cols) {
            break;
        }

        if grid[next_pos.row as usize][next_pos.col as usize] == '#' {
            current_direction = get_next_direction(current_direction);
            continue;
        }

        visited.insert(next_pos);
        current_pos = next_pos;
    }
}

fn creates_loop(
    grid: &[Vec<char>],
    initial_direction: char,
    start_pos: Position,
    obstacle_pos: Position,
) -> bool {
    let rows = grid.len() as i32;
    let cols = grid[0].len() as i32;

    let mut visited_states = HashSet::new();
    let mut current_pos = start_pos;
    let mut current_direction = initial_direction;

    let max_states = (rows * cols * 4) as usize; 

    loop {
        let next_pos = current_pos.next(current_direction);
        let state = State {
            pos: current_pos,
            direction: current_direction,
        };

        if !visited_states.insert(state) {
            return true;
        }

        if visited_states.len() > max_states {
            return false; 
        }

        if !next_pos.is_valid(rows, cols) {
            return false;
        }

        if next_pos == obstacle_pos || grid[next_pos.row as usize][next_pos.col as usize] == '#' {
            current_direction = get_next_direction(current_direction);
            continue;
        }

        current_pos = next_pos;
    }
}

fn part1() {
    let grid = build_grid("day06/input.txt");
    let (direction, pos) = find_guard(&grid);
    let mut visited = HashSet::new();
    visited.insert(pos);

    simulate_movement(&grid, direction, pos, &mut visited);
    println!("Part 1: {}", visited.len());
}

fn part2() {
    let grid = build_grid("day06/input.txt");
    let (initial_direction, initial_pos) = find_guard(&grid);
    let rows = grid.len();
    let cols = grid[0].len();

    let total = (0..rows)
        .flat_map(|row| (0..cols).map(move |col| (row, col)))
        .filter(|&(row, col)| {
            let pos = Position {
                row: row as i32,
                col: col as i32,
            };
            pos != initial_pos
                && grid[row][col] != '#'
                && creates_loop(&grid, initial_direction, initial_pos, pos)
        })
        .count();

    println!("Part 2: {}", total);
}

fn main() {
    part1();
    part2();
}
